### PR TITLE
[WIP] Finish partner, student hero, event and sub-event page 

### PR DESCRIPTION
**Pull request by khoainx**
@nguyenductoan please review and give me your opinion!

### CODE REVIEW

**Comment by khoainx** [comment]
@lhhieu89 please review my code and give me your opinion!

File: .byebug_history [hunk 1]
```diff
@@ -0,0 +1,95 @@
+continue
+@past_events.count
+@psat_events.count
+@comming_events.count
```
Comment (MikeHoang) [inline .byebug_history]: Please remove this file

File: app/controllers/admin/dashboard_controller.rb [hunk 2]
```diff
@@ -10,4 +10,15 @@ def set_partner
   def set_student_hero
     @heroes = StudentHero.all
   end
+  def all_notifications
+    @notifications = Notification.where(to_user: current_user.id).paginate(:page => params[:page], :per_page => 5)
+    @notification_of_user = Notification.all.where("to_user=?",current_user.id).where("status=?",nil)
```
Comment (MikeHoang) [inline app/controllers/admin/dashboard_controller.rb]: You should put it on Model

File: app/controllers/admin/partners_controller.rb [hunk 3]
```diff
@@ -39,6 +41,8 @@ def update
     else
       render 'edit'
     end
+    @noti = Notification.new(title: current_user.email + " has already updated your account", from_user: current_user.id, to_user: @partner.id, post_id: @partner.id, typeof:@partner.approve_type, notification_type: 1)
```
Comment (MikeHoang) [inline app/controllers/admin/partners_controller.rb]: Please put it on model and parse the last parameter into method

File: app/controllers/admin/pendings_controller.rb [hunk 4]
```diff
@@ -16,6 +16,13 @@ def destroy
   def update
     if @pending.update_attributes(approve_params)
       flash[:success]= "Update successfully"
+
+      #create notify when approve new partner to partner and partner who created
+      @partner_created_id = Notification.find_by(post_id: @pending.id).from_user
+      @noti = Notification.new(title: current_user.email + " has already accepted new partner:" + @pending.name, from_user: 1, to_user: @pending.id, post_id: @pending.id,typeof:@pending.approve_type, notification_type: 1)
```
Comment (MikeHoang) [inline app/controllers/admin/pendings_controller.rb]: Please put it on model too

File: app/controllers/event_editor/events_controller.rb [hunk 5]
```diff
@@ -41,14 +55,73 @@ def update
         format.json { render json: @event.errors, status: :unprocessable_entity }
       end
     end
+    # @users.each do |user|
+    #   if user.user_type == "partner" || user.user_type == "admin"
+    #     if current_user.user_type != user.user_type
+    #       @noti = Notification.new()
+    #       @noti.title = current_user.email + " has already updated Event"
+    #       @noti.from_user = current_user.id
+    #       @noti.to_user = user.id
```
Comment (MikeHoang) [inline app/controllers/event_editor/events_controller.rb]: Do you need this code anymore? If not, remove it

File: app/controllers/event_editor/events_controller.rb [hunk 6]
```diff
@@ -28,10 +29,23 @@ def create
         format.json { render json: @event.errors, status: :unprocessable_entity }
       end
     end
+    # create Notification when submitform
+
+    @users.each do |user|
+      if user.user_type == "partner" || user.user_type == "admin"
```
Comment (MikeHoang) [inline app/controllers/event_editor/events_controller.rb]: you can use enum to replace this.

File: app/controllers/event_editor/events_controller.rb [hunk 7]
```diff
@@ -41,14 +55,73 @@ def update
...
+    #       @noti.to_user = user.id
+    #       @noti.post_id = @event.id
+    #       @noti.typeof = "Event"
+    #       @noti.notification_type = 1
+    #       @noti.save!
+    #     end
+    #   end
+    # end
+    @users.each do |user|
+      if user.user_type == "partner" || user.user_type == "admin"
```
Comment (MikeHoang) [inline app/controllers/event_editor/events_controller.rb]: you can use enum to replace this.

File: app/controllers/event_editor/events_controller.rb [hunk 8]
```diff
@@ -41,14 +55,73 @@ def update
...
+        end
+      end
+    end
   end
-  def destroy 
-    @event.destroy 
+  def destroy
+    @users = User.all
+    @users.each do |user|
+      if user.user_type == "partner" || user.user_type == "admin"
```
Comment (MikeHoang) [inline app/controllers/event_editor/events_controller.rb]: you can use enum to replace this.

**Comment by khoainx** [comment]
@lhhieu89 please review my code and give me your opinion!

File: Gemfile [hunk 9]
```diff
@@ -28,7 +28,9 @@ gem 'slim-rails'
 gem 'fog'
 gem 'aws-sdk', '< 2'
 gem 'devise'
-
+## Gemfile for Rails 3+, Sinatra, and Merb
+gem 'will_paginate', '~> 3.0.6'
+gem 'will_paginate-bootstrap'
```
Comment (lhhieu89) [inline Gemfile]: Please remove 2 gem above and use `kaminari`, it's already installed

File: app/assets/javascripts/notifications.coffee [hunk 10]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (lhhieu89) [inline app/assets/javascripts/notifications.coffee]: Remove this file

File: app/controllers/admin/dashboard_controller.rb (hunk 2 above)
Comment (lhhieu89) [inline app/controllers/admin/dashboard_controller.rb]: Please use Rails Association

File: app/assets/javascripts/blogpage.coffee [hunk 11]
```diff
@@ -1,3 +0,0 @@
-# Place all the behaviors and hooks related to the matching controller here.
-# All this logic will automatically be available in application.js.
-# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (minhquan4080) [inline app/assets/javascripts/blogpage.coffee]: If don't use. Please remove it.

File: app/assets/javascripts/homepage.coffee [hunk 12]
```diff
@@ -1,3 +0,0 @@
-# Place all the behaviors and hooks related to the matching controller here.
-# All this logic will automatically be available in application.js.
-# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (minhquan4080) [inline app/assets/javascripts/homepage.coffee]: If don't use. Please remove it.

File: app/assets/javascripts/faqs.coffee [hunk 13]
```diff
@@ -1,3 +0,0 @@
-# Place all the behaviors and hooks related to the matching controller here.
-# All this logic will automatically be available in application.js.
-# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (minhquan4080) [inline app/assets/javascripts/faqs.coffee]: If don't use. Please remove it.

File: app/assets/javascripts/partner.coffee [hunk 14]
```diff
@@ -1,3 +0,0 @@
-# Place all the behaviors and hooks related to the matching controller here.
-# All this logic will automatically be available in application.js.
-# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (minhquan4080) [inline app/assets/javascripts/partner.coffee]: If don't use. Please remove it.

File: app/assets/javascripts/promotion.coffee [hunk 15]
```diff
@@ -1,3 +0,0 @@
-# Place all the behaviors and hooks related to the matching controller here.
-# All this logic will automatically be available in application.js.
-# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (minhquan4080) [inline app/assets/javascripts/promotion.coffee]: If don't use. Please remove it.

File: app/assets/javascripts/registrations_controller.coffee [hunk 16]
```diff
@@ -1,3 +0,0 @@
-# Place all the behaviors and hooks related to the matching controller here.
-# All this logic will automatically be available in application.js.
-# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (minhquan4080) [inline app/assets/javascripts/registrations_controller.coffee]: If don't use. Please remove it.

File: app/controllers/event_editor/events_controller.rb [hunk 17]
```diff
@@ -41,14 +55,59 @@ def update
         format.json { render json: @event.errors, status: :unprocessable_entity }
       end
     end
+    @users.each do |user|
+      if user.partner? || user.admin?
+        if current_user.user_type != user.user_type
+          @noti = Notification.new(title: current_user.email + " has already updated Event",from_user: current_user.id, to_user: user.id, post_id: @event.id, notification_type: 1, typeof: "Event")
+          @noti.save!
```
Comment (minhquan4080) [inline app/controllers/event_editor/events_controller.rb]: change  current_user.email + " has already updated Event" to  "#{current_user.email} has already updated Event"

File: app/controllers/event_editor/events_controller.rb [hunk 18]
```diff
@@ -41,14 +55,59 @@ def update
...
+    end
   end
-  def destroy 
-    @event.destroy 
+  def destroy
+    @users = User.all
+    @users.each do |user|
+      if user.partner? || user.admin?
+        if current_user.user_type != user.user_type
+          @noti = Notification.new(title: current_user.email + " has already updated Event",from_user: current_user.id, to_user: user.id, notification_type: 1, typeof: "Event")
```
Comment (minhquan4080) [inline app/controllers/event_editor/events_controller.rb]: change  : current_user.email + " has already updated Event" 
to : "#{current_user.email}  has already updated Event"

File: app/controllers/partner/promotions_controller.rb [hunk 19]
```diff
@@ -43,6 +43,9 @@ def create
       flash[:success] = "wait for admin approved!"
       render 'new'
     end
+     # create Notification when create parter from partner
+    @noti = Notification.new(title: current_user.email + " has already created new partner and waiting for approve", from_user: current_user.id, to_user: 1, post_id: @partner.id, typeof:@partner.approve_type, notification_type: 0)
```
Comment (minhquan4080) [inline app/controllers/partner/promotions_controller.rb]: change : current_user.email + " has already created new partner and waiting for approve"
to : "#{current_user.email} has already created new partner and waiting for approve"

File: app/controllers/partner/students_controller.rb [hunk 20]
```diff
@@ -70,9 +70,13 @@ def create
     @student.transac[:commission] = @student.transac.total_cost()*0.1
     if @student.save
       redirect_to partner_students_path
-    else 
+    else
       render 'new'
     end
+    #create notification when create student from partner
+    @noti = Notification.new(title:  current_user.email + " has already created user type: " + @student.user_type, from_user: current_user[:id], to_user: 1, post_id: @student.id, typeof: @student.user_type.capitalize, notification_type: 0)
```
Comment (minhquan4080) [inline app/controllers/partner/students_controller.rb]: change :  current_user.email + " has already created user type: " + @student.user_type
to : "#{current_user.email} has already created user type: #{@student.user_type}"

File: app/controllers/registrations_controller.rb [hunk 21]
```diff
@@ -21,6 +21,10 @@ def create
     end
     build_resource(tp)
     resource.save
+    # create Notification when submitform
+    @noti = Notification.new(title:  tp[:email] + " has already created user type: " + tp[:user_type], from_user: resource.id, to_user: 1, post_id: resource.id, typeof: tp[:user_type], notification_type: 0)
```
Comment (minhquan4080) [inline app/controllers/registrations_controller.rb]: change : tp[:email] + " has already created user type: " + tp[:user_type]
to : "#{tp[:email]}  has already created user type: #{tp[:user_type]}"

**Comment by minhquan4080** [comment]
@khoainx  please check my comment and fix conflics

**Comment by lhhieu89** [comment]
@khoainx : Please fix the conflict on your branch

**Pull request merged**